File: solutions/queue/ecommerce_order_log_problem/approach.js
class Record {
  constructor(N) {
    this.capacity = N;
    this.log = [];
    this.front = 0;
    this.rare = 0;
  }

  record(order_id) {
    if (this.log.length < this.capacity) {
      this.log.push(order_id);
      this.rare++;
    } else {
      this.rare = (this.rare + 1) % this.capacity;
      this.front = (this.front + 1) % this.capacity;
      this.log[this.rare] = order_id;
    }
  }

  get_last(index) {
    return this.log[(this.front + index - 1) % this.capacity];
  }
}

r = new Record(5);
r.record(102);
r.record(103);
r.record(104);
r.record(105);
r.record(106);
r.record(107);

console.log(r.get_last(3));

File: solutions/queue/ecommerce_order_log_problem/approach.py
class Record:
    def __init__(self,N) -> None:
        self.log = []
        self.capacity = N

    def record(self,order_id):
        if len(self.log) < self.capacity:
            self.log.append(order_id)
        else:
            self.log.pop(0)
            self.log.append(order_id)
    
    def get_last(self,index):
        return self.log[-index]
    
r = Record(5)
r.record(102)
r.record(103)
r.record(104)
r.record(105)
r.record(106)
r.record(107)

print(r.get_last(3))